File: backend/src/models/Download.ts
import mongoose, { Schema, Document } from 'mongoose';
import type { DownloadJob, VideoInfo } from '../types';

export interface IDownload extends Document, Omit<DownloadJob, '_id'> {}

const VideoInfoSchema = new Schema<VideoInfo>(
  {
    id: { type: String, required: true },
    title: { type: String, required: true },
    thumbnail: { type: String, required: true },
    duration: { type: Number, required: true },
    fileSize: { type: String },
    quality: { type: String },
  },
  { _id: false }
);

const DownloadSchema = new Schema<IDownload>(
  {
    url: {
      type: String,
      required: true,
      trim: true,
    },
    jobId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    status: {
      type: String,
      enum: ['queued', 'processing', 'completed', 'failed'],
      default: 'queued',
      required: true,
    },
    progress: {
      type: Number,
      min: 0,
      max: 100,
      default: 0,
    },
    videoInfo: {
      type: VideoInfoSchema,
    },
    downloadUrl: {
      type: String,
    },
    error: {
      type: String,
    },
  },
  {
    timestamps: true,
  }
);

// Index for cleanup queries
DownloadSchema.index({ createdAt: 1 });
DownloadSchema.index({ status: 1 });

export default mongoose.model<IDownload>('Download', DownloadSchema);

File: app/models/download.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from enum import Enum


class DownloadStatus(str, Enum):
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class VideoInfo(BaseModel):
    id: str
    title: str
    thumbnail: str
    duration: int
    file_size: Optional[str] = None
    quality: Optional[str] = None


class Download(BaseModel):
    job_id: str = Field(alias="jobId")
    url: str
    status: DownloadStatus = DownloadStatus.QUEUED
    progress: int = 0
    video_info: Optional[VideoInfo] = Field(None, alias="videoInfo")
    download_url: Optional[str] = Field(None, alias="downloadUrl")
    error: Optional[str] = None
    user_id: Optional[str] = Field(None, alias="userId")
    created_at: datetime = Field(default_factory=datetime.utcnow, alias="createdAt")
    updated_at: datetime = Field(default_factory=datetime.utcnow, alias="updatedAt")

    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class DownloadResponse(BaseModel):
    job_id: str = Field(alias="jobId")
    status: DownloadStatus
    progress: int = 0
    video_info: Optional[VideoInfo] = Field(None, alias="videoInfo")
    download_url: Optional[str] = Field(None, alias="downloadUrl")
    error: Optional[str] = None

    class Config:
        populate_by_name = True
